### PR TITLE
add orgs option to github extsvc config

### LINKED ISSUE #4036: GitHub external service config option for organizations
Currently, the recommended way to add all repos in a GitHub org to Sourcegraph is to use a repositoryQuery, but GitHub search API returns 1k repos max. We should add a config option for orgs in the GitHub external service config.

### PR DESCRIPTION
**Pull request by kzh**
This fixes #4036. This change grants users an alternative and safer way to add all the repositories from a GitHub organization to Sourcegraph using the `orgs` config option.

The previous method `repositoryQuery` used the GitHub search api which has a small rate limit so you would run into issues with incomplete results.

Config:
Screen Shot 2019-05-30 at 10 29 52 AM
Result:
Screen Shot 2019-05-30 at 10 30 31 AM

### CODE REVIEW

**Comment by beyang** [comment]
Add an entry to `CHANGELOG.md` describing the change.

File: pkg/extsvc/github/repos.go [hunk 1]
```diff
@@ -406,6 +406,24 @@ func (c *Client) ListUserRepositories(ctx context.Context, page int) (repos []*R
 	return repos, len(repos) > 0, 1, nil
 }
 
+// ListOrgRepositories lists GitHub repositories from the specified organization.
+// org is the name of the organization. page is the page of results to return.
+// Pages are 1-indexed (so the first call should be for page 1).
+func (c *Client) ListOrgRepositories(ctx context.Context, org string, page int) (repos []*Repository, hasNextPage bool, rateLimitCost int, err error) {
```
Comment (beyang) [inline pkg/extsvc/github/repos.go]: Add a unit test for this method. See https://sourcegraph.com/github.com/sourcegraph/sourcegraph@master/-/blob/pkg/extsvc/github/repos_test.go

File: pkg/extsvc/github/repos.go [hunk 2]
```diff
@@ -406,6 +406,24 @@ func (c *Client) ListUserRepositories(ctx context.Context, page int) (repos []*R
...
+	var restRepos []restRepository
+	path := fmt.Sprintf("orgs/%s/repos?sort=pushed&page=%d&per_page=100", org, page)
+	if err := c.requestGet(ctx, "", path, &restRepos); err != nil {
```
Comment (beyang) [inline pkg/extsvc/github/repos.go]: `requestGet` takes a token as its second argument. If we pass the empty token here, does this mean that we won't be fetching any of the organization's private repositories?

File: pkg/extsvc/github/repos.go [hunk 3]
```diff
@@ -406,6 +406,24 @@ func (c *Client) ListUserRepositories(ctx context.Context, page int) (repos []*R
...
+		return nil, false, 1, err
+	}
+	repos = make([]*Repository, 0, len(restRepos))
+	for _, restRepo := range restRepos {
+		repos = append(repos, convertRestRepo(restRepo))
+	}
+	// 🚨 SECURITY: must forward token here to ensure caching by token
+	c.addRepositoriesToCache("", repos)
```
Comment (beyang) [inline pkg/extsvc/github/repos.go]: If we *do* pass a token to `requestGet` above, make sure the pass the same token to `addRepositoriesToCache`.

File: schema/github.schema.json [hunk 4]
```diff
@@ -42,6 +42,12 @@
       "items": { "type": "string", "pattern": "^[\\w-]+/[\\w.-]+$" },
       "examples": [["owner/name"], ["kubernetes/kubernetes", "golang/go", "facebook/react"]]
     },
+    "orgs": {
+      "description": "An array of organization \"org name\" strings specifying which GitHub or GitHub Enterprise organizations to mirror all repositories from on Sourcegraph.",
```
Comment (beyang) [inline schema/github.schema.json]:
```suggestion
      "description": "An array of organization names identifying GitHub organizations whose repositories should be mirrored on Sourcegraph.",
```

File: web/src/site-admin/externalServices.tsx [hunk 5]
```diff
@@ -220,6 +220,11 @@ export const GITHUB_EXTERNAL_SERVICE: ExternalServiceKindMetadata = {
   //   "<owner>/<repository>"
   // ],
 
+  // orgs: Explicit list of organizations to select all repositories from
```
Comment (beyang) [inline web/src/site-admin/externalServices.tsx]:
```suggestion
  // orgs: List of organizations whose repositories should be selected
```

Comment (beyang) [reply web/src/site-admin/externalServices.tsx]: Move this in between the `repositoryQuery` and `repos` field.

File: cmd/repo-updater/repos/github.go [hunk 6]
```diff
@@ -310,6 +310,46 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
 		}
 	}
 
+	for _, org := range s.config.Orgs {
```
Comment (beyang) [inline cmd/repo-updater/repos/github.go]: The `listAllRepositories` method is quite large. It'd be best to split this up into 3 methods, 1 for `repositoryQuery`, 1 for `orgs`, 1 for `repos`.

Also, enhance/add unit tests for the `orgs` case to [TestGithubSource_ListRepos](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/github_test.go#L30:6)

**Comment by beyang** [comment]
@keegancsmith @tsenart @mrnugget can one of you review the repo-updater part of this change?

**Comment by keegancsmith** [comment]
Nice. I haven't looked at the code yet (will review tomorrow) but what do you think about adding a new config field (`org`) vs special casing `repositoryQuery`? The latter will make existing configs just work nicely. However, the former is likely more robust. cc @sourcegraph/core-services 

@beyang FYI we listed this as something we would work on in our 3.5 tracking issue. I wasn't aware https://github.com/sourcegraph/sourcegraph/issues/4036 existed. Do you have any other issues for 3.5 which core services cares about which I may have missed?

File: pkg/extsvc/github/repos.go (hunk 2 above)
Comment (kzh) [reply pkg/extsvc/github/repos.go]: The second argument is for explicitly using a different token from the one provided in the config. When an empty string is passed in, `requestGet` will fallback and opt for the config token stored within `Client`.

It should be grabbing all the repositories (public or private) of an organization.

```go
func (c *Client) do(ctx context.Context, token string, req *http.Request, result interface{}) (err error) {
	...
	if token != "" {
		req.Header.Set("Authorization", "bearer "+token)
	} else if c.defaultToken != "" {
		req.Header.Set("Authorization", "bearer "+c.defaultToken)
	}
        ...
}
```

```go
func NewClient(apiURL *url.URL, defaultToken string, cli httpcli.Doer) *Client {
        ...
	return &Client{
		...
		defaultToken: defaultToken,
		...
	}
}
```

**Comment by beyang** [comment]
> what do you think about adding a new config field (`org`) vs special casing `repositoryQuery`?

Extending `repositoryQuery` would be nice for the following reasons:
* One fewer config option, so config is more succinct.
* One config option for selecting sets of repos.

Adding `orgs` would be nice for the following reasons:
* We already have the `repos` field, and this would mirror that.
* It'd be more explicit, so we wouldn't have to explain the `org:$ORG` syntax in the docs. We'd also avoid any confusion with the `org:` keyword in the more general `repositoryQuery` syntax

Any other tradeoffs I missed? I have a slight preference toward `orgs`, because it is more explicit. If you feel strongly the other way, however, I'd rather go with your preference than discuss further.

If we do end up sticking with `orgs`, @kzh can you make the following additional edits:
* Update the [docs for `exclude`](https://docs.sourcegraph.com/admin/external_service/github#selecting-repositories-for-code-search) to mention that it takes precedence over the `repos`, `repositoryQuery`, *and* `orgs` fields.

**Comment by kzh** [comment]
Got it. I'm currently refactoring the code (mainly breaking up `listAllRepositories` into smaller methods) and will wrap up early tomorrow.

**Comment by mrnugget** [comment]
@beyang @kzh We just discussed this question of `orgs` vs. special-case-`org:`-queries in @sourcegraph/core-services and think that the best approach would be to do both:
1. add the `orgs` field by landing this PR and then...
2. do a follow-up PR that uses the new methods added here to fix https://github.com/sourcegraph/sourcegraph/issues/4036 by special-casing the absolute simplest version of `org:` queries (just `"org:<org-name>"`) to use the `ListOrgRepositories` methods.

What do you think?

**Comment by keegancsmith** [comment]
FYI the motivation for special casing in repositoryQuery as well is so existing customers immediately benefit. Bonus points for supporting queries that match the regex `(org:(\S+)\s*)+`.

File: CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,8 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Added
 
+- The `orgs` setting in [GitHub external service config](https://docs.sourcegraph.com/admin/external_service/github) allows admins to select all repositories from the specified organizations to be synced.
+
```
Comment (tsenart) [inline CHANGELOG.md]: That documentation page should be updated too: https://github.com/sourcegraph/sourcegraph/blob/master/doc/admin/external_service/github.md#selecting-repositories-for-code-search

File: cmd/repo-updater/repos/github.go [hunk 8]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
 	return s.exclude[strings.ToLower(r.NameWithOwner)] || s.exclude[r.ID]
 }
 
-func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repository, error) {
-	set := make(map[int64]*github.Repository)
-	errs := new(multierror.Error)
+// RepositoryPager is a function that takes in a page number and returns
+type RepositoryPager func(page int) (repos []*github.Repository, hasNextPage bool, wait int, err error)
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Does this type need to be exported?

File: cmd/repo-updater/repos/github.go [hunk 9]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
 
-	for _, repositoryQuery := range s.config.RepositoryQuery {
-		switch repositoryQuery {
-		case "public":
-			if s.githubDotCom {
-				errs = multierror.Append(errs, errors.New(`unsupported configuration "public" for "repositoryQuery" for github.com`))
+// exhaustRepositoryPager takes in a RepositoryPager and accumulates all the repositories from every page.
+func (s *GithubSource) exhaustRepositoryPager(ctx context.Context, pager RepositoryPager) (set map[int64]*github.Repository, err error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
func (s *GithubSource) paginate(ctx context.Context, pager RepositoryPager) (set map[int64]*github.Repository, err error) {
```

File: cmd/repo-updater/repos/github.go [hunk 10]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+		for _, r := range pageRepos {
+			set[r.DatabaseID] = r
+		}
+
+		if hasNextPage && wait > 0 {
+			time.Sleep(s.client.RateLimit.RecommendedWaitForBackgroundOp(wait))
+		}
+	}
+
+	return
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: This method is too long for naked returns to be an advantage for readability. Please make returns explicit. For more context:

> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: cmd/repo-updater/repos/github.go (hunk 8 above)
Comment (tsenart) [reply cmd/repo-updater/repos/github.go]: What are the semantics of the `wait` return argument?

File: cmd/repo-updater/repos/github.go [hunk 11]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+}
+
+// listOrgRepositories handles the `orgs` option.
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Can we describe what each of these methods does in addition to what part of the external service configuration it handles?

File: cmd/repo-updater/repos/github.go [hunk 12]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+				"rateLimitReset", rateLimitReset,
+				"retryAfter", rateLimitRetry,
+			)
+			return
+		},
+	)
+}
+
+// listSelectedRepositories handles the `repos` option.
+func (s *GithubSource) listSelectedRepositories(ctx context.Context, names []string) (set map[int64]*github.Repository, err error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
func (s *GithubSource) listRepos(ctx context.Context, names []string) (set map[int64]*github.Repository, err error) {
```

File: cmd/repo-updater/repos/github.go [hunk 13]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
-				}
-				log15.Debug("github sync public", "repos", len(repos), "error", err)
-				for _, r := range repos {
-					set[r.DatabaseID] = r
-					if sinceRepoID < r.DatabaseID {
-						sinceRepoID = r.DatabaseID
-					}
-				}
+// listQueryRepositories handles the `repositoryQuery` option.
+func (s *GithubSource) listQueryRepositories(ctx context.Context, query string) (set map[int64]*github.Repository, err error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
func (s *GithubSource) listRepositoryQuery(ctx context.Context, query string) (set map[int64]*github.Repository, err error) {
```

File: cmd/repo-updater/repos/github.go [hunk 14]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+		if hasNextPage && wait > 0 {
+			time.Sleep(s.client.RateLimit.RecommendedWaitForBackgroundOp(wait))
+		}
+	}
+
+	return
+}
+
+// listOrgRepositories handles the `orgs` option.
+func (s *GithubSource) listOrgRepositories(ctx context.Context, org string) (map[int64]*github.Repository, error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
func (s *GithubSource) listOrg(ctx context.Context, org string) (map[int64]*github.Repository, error) {
```

File: cmd/repo-updater/repos/github.go [hunk 15]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+			if err != nil {
+				err = errors.Wrapf(err, "failed to list public repositories: sinceRepoID=%d", sinceRepoID)
+				return
+			}
+			if len(repos) == 0 {
+				return
+			}
+			log15.Debug("github sync public", "repos", len(repos), "error", err)
+			for _, r := range repos {
+				fmt.Println(r.NameWithOwner)
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Leftover.

File: cmd/repo-updater/repos/github.go (hunk 12 above)
Comment (tsenart) [reply cmd/repo-updater/repos/github.go]: I believe that `s.fetchAllRepositoriesInBatches(ctx, set)` would be best attempted here first, instead of keeping that fallback logic where it currently is.

File: cmd/repo-updater/repos/github.go [hunk 16]
```diff
@@ -233,83 +307,82 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
...
+		)
+	}
+	return
+}
+
+func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repository, error) {
+	set := make(map[int64]*github.Repository)
+	errs := new(multierror.Error)
+
+	// repositoryQuery
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
```

File: cmd/repo-updater/repos/github.go [hunk 17]
```diff
@@ -233,83 +307,82 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
...
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+		for id, r := range repos {
+			set[id] = r
 		}
 	}
 
+	// repos
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
```

File: cmd/repo-updater/repos/github.go [hunk 18]
```diff
@@ -320,37 +393,30 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
...
-				break
-			}
+		for id, r := range repos {
+			set[id] = r
+		}
+	}
 
-			log15.Debug("github sync: GetRepository", "repo", repo.NameWithOwner)
-			set[repo.DatabaseID] = repo
+	// orgs
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
```

File: cmd/repo-updater/repos/github.go [hunk 19]
```diff
@@ -320,37 +393,30 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
...
+			continue
+		}
 
-			time.Sleep(s.client.RateLimit.RecommendedWaitForBackgroundOp(1)) // 0-duration sleep unless nearing rate limit exhaustion
+		for id, r := range repos {
+			set[id] = r
 		}
 	}
 
+	// exclude
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]:
```suggestion
```

File: cmd/repo-updater/repos/github.go [hunk 20]
```diff
@@ -174,56 +174,130 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+			for _, r := range repos {
+				fmt.Println(r.NameWithOwner)
+				set[r.DatabaseID] = r
+				if sinceRepoID < r.DatabaseID {
+					sinceRepoID = r.DatabaseID
 				}
+			}
+		}
+	case "affiliated":
+		set, err = s.exhaustRepositoryPager(
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Maybe worth extracting this to its own method?

File: cmd/repo-updater/repos/github.go [hunk 21]
```diff
@@ -233,83 +307,82 @@ func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repos
...
-				reposPage, err := s.searchClient.ListRepositoriesForSearch(ctx, repositoryQuery, page)
+				return
+			},
+		)
+	case "none":
+		// nothing
+	default:
+		// Run the query as a GitHub advanced repository search
+		// (https://github.com/search/advanced).
+		set, err = s.exhaustRepositoryPager(
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Here it'd also be good to extract to its own method.

File: pkg/extsvc/github/repos_test.go [hunk 22]
```diff
@@ -361,6 +361,56 @@ func TestClient_GetRepositoryByNodeID_nonexistent(t *testing.T) {
...
+			NameWithOwner: "o/r",
+			Description:   "d",
+			URL:           "https://github.example.com/o/r",
+			IsFork:        true,
+		},
+		{
+			ID:            "j",
+			NameWithOwner: "a/b",
+			Description:   "c",
+			URL:           "https://github.example.com/a/b",
```
Comment (mrnugget) [inline pkg/extsvc/github/repos_test.go]: I admit, this is the ultimate nitpick, but why do the repositories here belong to two different organisations? Does this reflect what could be returned?

I personally like when the data/mocks used in a test setup reflect real-world data and make clear what the assertion will be about, or at least help in understanding it. In this case, I was reading through through the test and thought "Ah, I bet the assertion is that only one of the two repositories is returned".

Comment (mrnugget) [reply pkg/extsvc/github/repos_test.go]: Adding to that: below you use `"sourcegraph"` as the org in the call, which is then the third org name in this test

File: pkg/extsvc/github/repos.go (hunk 3 above)
Comment (kzh) [reply pkg/extsvc/github/repos.go]: The same interaction happens here as with `requestGet`.

File: cmd/repo-updater/repos/github.go (hunk 8 above)
Comment (kzh) [reply cmd/repo-updater/repos/github.go]: `wait` is used to determine the delay before re-invoking `pager` for the next page. This is to ensure we don't hit rate limits. For many of the pager functions, we return `rateLimitCost` as `wait`.

```go
func (s *GithubSource) paginate(ctx context.Context, pager repositoryPager) (map[int64]*github.Repository, error) {
        ...
	hasNextPage := true
	for page := 1; hasNextPage; page++ {
		...
		if hasNextPage && wait > 0 {
			time.Sleep(s.client.RateLimit.RecommendedWaitForBackgroundOp(wait))
		}
	}
        ...
}
```

File: pkg/extsvc/github/repos_test.go (hunk 22 above)
Comment (kzh) [reply pkg/extsvc/github/repos_test.go]: Ah, that makes sense. I changed all the organization names to `o`.

File: cmd/repo-updater/repos/github.go (hunk 8 above)
Comment (tsenart) [reply cmd/repo-updater/repos/github.go]: So, it's not as much the wait time, but more the cost, right? I'd adjust the name accordingly.

File: pkg/extsvc/github/repos.go [hunk 23]
```diff
@@ -479,6 +479,24 @@ func (c *Client) ListUserRepositories(ctx context.Context, page int) (repos []*R
...
+	return repos, len(repos) > 0, 1, nil
```
Comment (tsenart) [inline pkg/extsvc/github/repos.go]: I know this was copied from another method, but I don't think this method of determining if there is a next page in the result set is correct. The GitHub API provides explicit pagination information in the response, so we should use that. Perhaps you could follow up with a PR for that.

File: cmd/repo-updater/repos/github.go [hunk 24]
```diff
@@ -174,180 +174,275 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
-					// less than 5 requests left.
-					remaining, _, retryAfter, ok := s.searchClient.RateLimit.Get()
-					if retryAfter > 0 || (ok && remaining < 5) {
-						time.Sleep(s.searchClient.RateLimit.RecommendedWaitForBackgroundOp(1))
-					}
-				}
-			}
+		pageRepos, hasNextPage, wait, err := pager(page)
+		if err != nil {
+			break
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: It seems like we are no longer propagating errors to the caller. Was this change intentional?

File: cmd/repo-updater/repos/github.go [hunk 25]
```diff
@@ -174,180 +174,275 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
-	err := s.fetchAllRepositoriesInBatches(ctx, set)
-	if err != nil {
+	return set, nil
+}
+
+// listOrg handles the `org` config option.
+// It returns all the repositories belonging to the given organization
+// by hitting the /orgs/:org/repos endpoint.
+func (s *GithubSource) listOrg(ctx context.Context, org string) (map[int64]*github.Repository, error) {
+	return s.paginate(
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Nitpick suggestion for conciseness:

```go
// listOrg handles the `org` config option.
// It returns all the repositories belonging to the given organization
// by hitting the /orgs/:org/repos endpoint.
func (s *GithubSource) listOrg(ctx context.Context, org string) (map[int64]*github.Repository, error) {
	return s.paginate(ctx, func(page int) (repos []*github.Repository, hasNext bool, cost int, err error) {
		defer func() {
			remaining, reset, retry, _ := s.client.RateLimit.Get()
			log15.Debug(
				"github sync: ListOrgRepositories",
				"repos", len(repos),
				"rateLimitCost", cost,
				"rateLimitRemaining", remaining,
				"rateLimitReset", reset,
				"retryAfter", retry,
			)
		}()
		return s.client.ListOrgRepositories(ctx, org, page)
	})
}
```

File: cmd/repo-updater/repos/github.go [hunk 26]
```diff
@@ -174,180 +174,275 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+			)
+			return
+		},
+	)
+}
+
+// listRepos returns the valid repositories from the given list of repository names.
+// This is done by hitting the /repos/:owner/:name endpoint for each of the given
+// repository names.
+func (s *GithubSource) listRepos(ctx context.Context) (map[int64]*github.Repository, error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Similarly to what we do in the other `listXXX` methods, we should probably parameterise the list of repos to be listed here.

```suggestion
func (s *GithubSource) listRepos(ctx context.Context, repos ...string) (map[int64]*github.Repository, error) {
```

File: cmd/repo-updater/repos/github.go [hunk 27]
```diff
@@ -174,180 +174,275 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+}
+
+// listRepositoryQuery handles the `repositoryQuery` config option.
+// The supported keywords to select repositories are:
+// - `public`: public repositories (from endpoint: /repositories)
+// - `affiliated`: repositories affiliated with client token (from endpoint: /user/repos)
+// - `none`: disables `repositoryQuery`
+// Inputs other than these three keywords will be queried using
+// GitHub advanced repository search (endpoint: /search/repositories)
+func (s *GithubSource) listRepositoryQuery(ctx context.Context, query string) (map[int64]*github.Repository, error) {
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Very nice!

File: cmd/repo-updater/repos/github.go [hunk 28]
```diff
@@ -174,180 +174,275 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+		for id, r := range repos {
+			set[id] = r
+		}
+	}
+
+	{
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Is this block necessary?

File: pkg/extsvc/github/repos.go (hunk 23 above)
Comment (kzh) [reply pkg/extsvc/github/repos.go]: Ah, I see in the `Link` response header. I'll get to fixing that.

```
Link: <https://api.github.com/resource?page=2>; rel="next",
      <https://api.github.com/resource?page=5>; rel="last"
```

File: pkg/extsvc/github/repos.go (hunk 3 above)
Comment (kzh) [reply pkg/extsvc/github/repos.go]: Ah gotcha, now that I think about it, the caching here is negligible. There are only a few unlikely instances in which caching would be helpful here such as if there was a redundant `repos` repository (ex: sourcegraph/sourcegraph) configured that overlapped with a configured `orgs` organization (ex: sourcegraph).

It wouldn't make sense to cache the entire organization either, since the state of the organization is constantly changing (repos are constantly being added/deleted) while the information regarding a specific repository doesn't change much.

Q: Would this be true for the `ListRepositoriesForSearch` as well? I see that it also caches the repositories queried using GitHub advanced repository search.

Comment (tsenart) [reply pkg/extsvc/github/repos.go]: > Q: Would this be true for the ListRepositoriesForSearch as well? I see that it also caches the repositories queried using GitHub advanced repository search.

I'd say so.

**Comment by kzh** [comment]
@beyang @mrnugget @tsenart
Thanks for the review and kind words! These comments helped tremendously. The code should be nearing completion now aside from a few minor hiccups.

Following this commit, I'll submit another pull request removing the `orgs` config option in favor of using `org:<org name>` special casing in the `repositoryQuery` config option. It should be simple enough.

@mrnugget
I added `orgs` to the `ListRepo` unit test. If you could take a took, that would be great. I was also wondering if we were transitioning over to using github's graphql api in place of the rest api.
```go
assertAllReposListed := func(t testing.TB, rs Repos) {
	t.Helper()

	want := []string{
		"github.com/kubernetes/kubeadm",
		"github.com/kubernetes/kubectl",
		"github.com/kubernetes/kubernetes",
		"github.com/kubernetes/minikube",
		"github.com/sourcegraph/about",
		"github.com/sourcegraph/sourcegraph",
	}

	have := rs.Names()
	sort.Strings(have)

	if !reflect.DeepEqual(have, want) {
		t.Error(cmp.Diff(have, want))
	}
}
...
svc := &ExternalService{
	Kind: "GITHUB",
	Config: marshalJSON(t, &schema.GitHubConnection{
	Url:   "https://github.com",
	Token: os.Getenv("GITHUB_ACCESS_TOKEN"),
	Repos: []string{
		"sourcegraph/about",
		"sourcegraph/sourcegraph",
	},
	Orgs: []string{
		"kubernetes",
	},
})
```

Also:
Something peculiar that I noticed with Beyang was that the github advanced repository search results for some organizations (ex: [`org:sourcegraph`](https://github.com/search?q=org%3Asourcegraph)) returned a smaller number than the actual number of repositories that the organization (ex: [sourcegraph](https://github.com/sourcegraph/sourcegraph)) owned, including organizations with fewer than 1000 repositories. Does anyone have any insight on which repositories are being left out (possibly forks, archives, etc)?

File: cmd/repo-updater/repos/github.go [hunk 29]
```diff
@@ -174,180 +174,270 @@ func (s *GithubSource) excludes(r *github.Repository) bool {
...
+	return s.listSearch(ctx, query)
+}
+
+// listAllRepositories returns the repositories from the given `orgs`, `repos`, and
+// `repositoryQuery` config options excluding the ones specified by `exclude`.
+func (s *GithubSource) listAllRepositories(ctx context.Context) ([]*github.Repository, error) {
+	set := make(map[int64]*github.Repository)
+	errs := new(multierror.Error)
+
+	var list map[int64]*github.Repository
```
Comment (tsenart) [inline cmd/repo-updater/repos/github.go]: Is there a good reason to declare `list` and `err` here?

**Comment by mrnugget** [comment]
> @mrnugget
> I added `orgs` to the `ListRepos` unit test. If you can take a look, that would be great! Also, are we prioritizing github's graphql api over the rest api?

Looks good to me! But as you noticed in your `edit:` the requests/responses are auto-saved as fixtures. You can update the fixtures by running the tests with the `-update <name-of-test-fixture-folder>` flag, like this:

```
pushd cmd/repo-updater/repos && go test -run TestGithubSource_ListRepos ./... -update GITHUB-LIST-REPOS && popd
```

Now, if you're going to run this, I bet that your test will break since the kubernetes org has more than 4 repositories. So I suggest you use an org with a small number of repositories, for example: https://github.com/golang-migrate That has only one repository. But maybe even use the `gorilla` org (18 repos) to make sure that loading _multiple_  repos works. Or maybe there's another small-ish org that helps execute as many code paths as possible?

And to add another suggestion: maybe now's the time to make the test cases in `ListRepos` more configurable, so we can set a separate external-service-config and assertion per test case? That would make failures easier to debug (if for example the GraphQL code for `config.Repos` breaks then ideally the test case for your new `org:` feature shouldn't break, etc.)

Regarding GraphQL vs. REST: I don't think we've made a decision about this yet, but I moved to the GraphQL API for the batch-fetching because that's not possible in the REST API. @beyang had the same reason for adding the initial support for the GraphQL API, as I understand. 

One thing to consider when talking about these two APIs is what @keegancsmith & @tsenart mentioned to me: in the past some of the GraphQL features were only available on Github.com but not on GitHub Enterprise, which blocked the move to GraphQL. Maybe now's the time to re-evaluate that.

**Comment by mrnugget** [comment]
> Following this commit, I'll submit another pull request removing the `orgs` config option in favor of using `org:<org name>` special casing in the `repositoryQuery` config option. It should be simple enough.

Just to be sure: is that the decision we came up with? (cc @beyang)

Personally, I'd be fine with having the `orgs` option AND special-casing `repositoryQuery: ["org:foobar"]` to use the optimized fetching. Removing `orgs` is also fine with me, I just want to be sure that this is the result of a consciously decision, not a misunderstanding :)

**Comment by tsenart** [comment]
> Personally, I'd be fine with having the orgs option AND special-casing repositoryQuery: ["org:foobar"] to use the optimized fetching.

I also thought this was what we determined.

File: cmd/repo-updater/repos/github.go (hunk 29 above)
Comment (kzh) [reply cmd/repo-updater/repos/github.go]: This was my mistake. The declaration is not needed.

**Comment by kzh** [comment]
Oh sorry! I misunderstood. I'll add the special-casing in the next pull request without removing `orgs`. 

@mrnugget Got it. The `orgs` unit test is now isolated from the other tests.
